settings_users: Add role based filter to the active and deactivated user list page.

shubham-padia opened the pull request:
Fixes #18617.

The role filter is added to both active and deactivated user list. The original plan was to introduce a multi select checkbox dropdown for the roles, but since that component is not ready yet, we will use the dropdown component for the first iteration.
To accomodate multiple filters, we have used an approach similar to the one in recent_view_ui.js where we use dropdown callback function and on("input") event on text search to set the filter value in memory. The predicate functions uses these filters. See this [CZO chat](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Allow.20multiple.20filters.20in.20ListWidget) for more details.

There was an existing inactive PR linked to this issue, that PR was not used since it was introducing business logic in ListWidget directly and it would be easier to make this PR from scratch compared to using that one and expanding on it. 

I have not added any E2E tests, since there were none existing covering this area, let me know if you think E2E tests are needed. 

<details>
<summary>Previous implementation details (for history):</summary>
~~The PR introduces a hidden input field which contains the filter info in json, the ListWidget will use that hidden input field as the main input and the filterer function will run on that field's value. Storing the value in json, then serialising and deserialising feels bit of a hacky solution, but I'm not sure whether it's worth to refactor ListWidget right now to accommodate that requirement natively.~~
 
~~Ideally I would have had two different commits to add the filter to active user list and deactivated user list, but since so much of the code is shared, it was easier to do it in one commit.~~
</details>


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


## Screenshots

<details>
<summary>Open role filter dropdown on active users page:</summary>

![Screenshot 2024-04-05 at 10 51 15 PM](https://github.com/zulip/zulip/assets/13910561/5112a640-23a8-4f0f-8bcb-63753ee40f49)

![Screenshot 2024-04-05 at 10 51 25 PM](https://github.com/zulip/zulip/assets/13910561/f9832fd5-850c-4d04-9058-ceec891725a0)

</details>

<details>
<summary>Open role filter dropdown on deactivated users page:</summary>

![Screenshot 2024-04-05 at 10 49 40 PM](https://github.com/zulip/zulip/assets/13910561/47e2add2-c07a-4bff-ab95-f369107ab45c)

</details>


<details>
<summary>Member role filtered on active users page with text search `a`:</summary>

![Screenshot 2024-04-05 at 10 52 22 PM](https://github.com/zulip/zulip/assets/13910561/cb3128f4-7b6d-42ea-ac25-3bb291c93424)

</details>


<details>
<summary>Moderator role filtered on deactivated users page:</summary>

![Screenshot 2024-04-05 at 10 50 18 PM](https://github.com/zulip/zulip/assets/13910561/fa82cb53-11f1-4de7-9844-63f346035ce7)

</details>

## Self review checklist

<details>
<summary>Self-review checklist</summary>


- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! As requested in the PR template, please post screenshots of your changes in the PR description (including an open dropdown). It is very difficult to review a video.

shubham-padia commented:
> of your changes in the PR description (including an open dropdown). It is very difficult to review a video.

Got it. Updated the PR description.

alya commented:
Nice, those all look good!

Playing with the PR, I think we should change "No users match your current filter." -> "No users match your filters." I think the current logic for that looks fine (any situation other than when all users are shown); we don't need to customize it based on the details of how the two filtering options are configured.

alya commented:
Could we prevent the "Actions" header from sliding over to the right when there are no results? This is not a regression in this PR, so we can file a separate issue for it if that makes sense.

![CleanShot 2024-04-03 at 13 48 22](https://github.com/zulip/zulip/assets/2090066/0656d681-3994-4a79-ab41-46a994a5f6cc)
![CleanShot 2024-04-03 at 13 49 11](https://github.com/zulip/zulip/assets/2090066/505ebdb5-5562-46fd-8f1d-10cf97e88401)

alya commented:
No other feedback from my side. @sahil839 are you up for a review?

shubham-padia commented:
> I think we should change "No users match your current filter." -> "No users match your filters."

Pushed the changes.

<details>
<summary> Active user list page </summary>
<img width="1291" alt="Screenshot 2024-04-04 at 2 22 18 AM" src="https://github.com/zulip/zulip/assets/13910561/9b6663f0-2380-45c6-82e3-b2bbac0233fb">

</details>

<details>
<summary> Deactivated user list page </summary>
<img width="1291" alt="Screenshot 2024-04-04 at 2 22 29 AM" src="https://github.com/zulip/zulip/assets/13910561/292b52ac-cfbd-4033-b32a-c06369ff85e9">

</details>

shubham-padia commented:
> Could we prevent the "Actions" header from sliding over to the right when there are no results? This is not a regression in this PR, so we can file a separate issue for it if that makes sense.
> 
> ![CleanShot 2024-04-03 at 13 48 22](https://private-user-images.githubusercontent.com/2090066/319351416-0656d681-3994-4a79-ab41-46a994a5f6cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIxNzc5NzUsIm5iZiI6MTcxMjE3NzY3NSwicGF0aCI6Ii8yMDkwMDY2LzMxOTM1MTQxNi0wNjU2ZDY4MS0zOTk0LTRhNzktYWI0MS00NmE5OTRhNWY2Y2MucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDNUMjA1NDM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjE3MDIxNDU4MjlmNmEzYzBhNTIzZmYxNDMzMDQxYWM4MGE4Zjg4OTIwYWYzMGM4ZDY5MDkyOWVkNmQxMGFiZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.5R2Hf50bkzsagJkM6IE4qihAwMGfaXbxNhlnHk9t6eM) ![CleanShot 2024-04-03 at 13 49 11](https://private-user-images.githubusercontent.com/2090066/319351427-505ebdb5-5562-46fd-8f1d-10cf97e88401.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIxNzc5NzUsIm5iZiI6MTcxMjE3NzY3NSwicGF0aCI6Ii8yMDkwMDY2LzMxOTM1MTQyNy01MDVlYmRiNS01NTYyLTQ2ZmQtOGYxZC0xMGNmOTdlODg0MDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDNUMjA1NDM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjU2NGJiOTRiZDhiM2U1ZjgxMTExZmFjNDcyNDdhOGQ4Yzk0NTRiMWUzY2Y5NzFjOTQzZWI1ZmJjNDMwZTAyYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.0FnNyPDuKRpC56sIyvKxB5qGRE8zAolMHIDY8Mw6eak)

I've started a convo on this [CZO link](https://chat.zulip.org/#narrow/stream/9-issues/topic/Last.20column.20in.20table.20slides.20to.20the.20right/near/1774355), so we can clarify the expected behaviour before opening an issue.

amanagr commented on web/src/settings_users.js:
```diff
@@ -199,6 +224,22 @@ function human_info(person) {
+        default_id: "0",
+        tippy_props: {
+            placement: "bottom-start",
+            // Without 0, 0 offset, the dropdown list starts a few pixels
+            // below the input area leaving a small gap.
+            offset: [0, 0],
```
That's intentional.

amanagr replied:
We should modify all the dropdowns for any options we decide to change here. I do feel like that gap is a bit much there, so maybe we can just reduce that.

amanagr commented on web/src/settings_users.js:
```diff
@@ -199,6 +224,22 @@ function human_info(person) {
 
+function create_role_filter_dropdown($events_container) {
+    new dropdown_widget.DropdownWidget({
+        widget_name: "select_role",
```
We should use a more specific name here, like `user_list_select_user_role`.

amanagr replied:
Oh! We need to pass different widget_name for dropdown in users list and deactivated users list, otherwise it will create two widgets with the same id.

shubham-padia replied:
Fixed both

amanagr commented on web/styles/settings.css:
```diff
@@ -1886,6 +1886,23 @@ $option_title_width: 180px;
+        flex-wrap: wrap;
+
+        #select_role_widget {
+            /* The font size here is 0.1 em less than input.search which
+            is just beside this widget. We kept it less so that their
+            resultant height is same in the end. This widget was slightly
+            bigger otherwise. */
+            font-size: 0.9em;
```
I don't like modifying font-size to adjust height of the dropdown widget here. You can override styles for `.dropdown-widget-button` to adjust the height.

shubham-padia replied:
Fixed it

amanagr commented on web/src/people.ts:
```diff
@@ -1592,7 +1592,18 @@ export function matches_user_settings_search(person: User, value: string): boole
     return safe_lower(person.full_name).includes(value) || safe_lower(email).includes(value);
 }
 
+function filter_for_user_roles(persons: User[], codeString: string): User[] {
```
We use snake_case for variable names.

shubham-padia replied:
not relevant anymore since the code has changed

amanagr commented:
I am not sure I like the idea of using an element like this to parse the current filter:
```
<input type="hidden" class="filter_json" value='{"text_search": "", "role": "0"}' />
```

Maybe because it feels like indirect approach?

A cleaner solution might be to have keep the filters registered in list_widget the same, but modify the filter callback to account for the selected value in dropdown widget. Also, the dropdown widget callback for change in selected value should call the filter callback registered in list_widget with the new selected value.

sahil839 commented:
I also agree. Might be better to just update the list_widget code. We currently handle only single filter, so we might just update the list_widget code to pass a list of filters and might be useful for other tables as well in future.

shubham-padia commented:
@amanagr @sahil839 Thanks for the review, I agree. I'll think about how we can modify the ListWidget and put a proposal tomorrow on the frontend channel for the new filter api for ListWidget. I'll then update this PR with the code changes once we've agreed on an interface.

shubham-padia commented:
I've updated the PR with the implementation discussed at this [CZO chat](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Allow.20multiple.20filters.20in.20ListWidget). @amanagr PTAL when you get some time, thanks! 

I've updated the screenshots, but removed the screen captures since they were not required last time.

amanagr commented:
@shubham-padia thanks! LGTM!

timabbott commented on web/src/admin.js:
```diff
@@ -224,6 +224,9 @@ export function build_page() {
         realm_enable_guest_user_indicator: realm.realm_enable_guest_user_indicator,
+        active_user_list_dropdown_widget_name: settings_users.active_user_list_dropdown_widget_name,
+        deactivated_user_list_dropdown_widget_name:
+            settings_users.deactivated_user_list_dropdown_widget_name,
```
Just a small note that in https://github.com/zulip/zulip/issues/26949, we'll probably end up needing only one of these, but I like the naming scheme.

timabbott commented on web/templates/settings/user_list_admin.hbs:
```diff
@@ -5,7 +5,10 @@
         <h3>{{t "Users"}}</h3>
         <div class="alert-notification" id="user-field-status"></div>
-        <input type="text" class="search filter_text_input" placeholder="{{t 'Filter users' }}" aria-label="{{t 'Filter users' }}"/>
+        <div class="user_filters">
```
Small naming comment: We've been moving towards a convention of preferring to add `-s`  over `_`s in CSS class names; but not a high priority worth reworking anything over.

timabbott commented on web/src/settings_users.js:
```diff
@@ -437,9 +516,19 @@ function handle_edit_form($tbody) {
+        .on("input.list_widget_filter", function () {
+            add_value_to_filters(section, "text_search", this.value.toLocaleLowerCase());
+        });
+}
```
What does this block of code do? It looks like it pseudo-duplicates `set_up_event_handlers` in `ListWidget`...

timabbott replied:
Ahh, OK, we're not using that `ListWidget` feature; added a comment before merging to address this.

timabbott commented on web/src/settings_users.js:
```diff
@@ -94,6 +113,52 @@ function failed_listing_users() {
+        {unique_id: "0", name: $t({defaultMessage: "All roles"})},
+        ...Object.values(settings_config.user_role_values).map((user_role_value) => ({
+            unique_id: user_role_value.code.toString(),
+            name: user_role_value.description,
```
I guess we might need to add a `description_plural` field in `settings_config.user_role_values` if we want the items in the dropdown to be plural, e.g., "Admins".

timabbott commented on web/src/settings_users.js:
```diff
@@ -94,6 +113,52 @@ function failed_listing_users() {
+            offset: [0, 0],
+        },
+    }).setup();
+}
```
Having a `section` parameter and defining `role_selected_handler` inside this function might let us clean up `role_selected_handler` to not need the conditionals ... but I don't think it's worth reworking given that we intended to merge the panels anyway.

timabbott commented:
Merged after making the tweak noted in https://github.com/zulip/zulip/pull/29597#discussion_r1560036497:
``` diff
diff --git a/web/src/settings_users.js b/web/src/settings_users.js
index 7b97d8d4c6..6e3cb8ca3d 100644
--- a/web/src/settings_users.js
+++ b/web/src/settings_users.js
@@ -517,6 +517,10 @@ function handle_edit_form($tbody) {
 }
 
 function handle_filter_change($tbody, section) {
+    // This duplicates the built-in search filter live-update logic in
+    // ListWidget for the input.list_widget_filter event type, but we
+    // can't use that, because we're also filtering on Role with our
+    // custom predicate.
     $tbody
         .closest(".settings-section")
         .find(".search")
```

Thanks @shubham-padia!

https://github.com/zulip/zulip/issues/26949 is probably the next thing we should do in this part of settings, since it'll simplify the logic quite a bit.

alya commented:
I'm working on user documentation for this feature, which reminded me that a major use case is being able to find admins in your org. @shubham-padia Would you be up for putting up a quick PR that reverses the role order in the dropdown? So it would be: All roles, Owner, Administrator, etc.

shubham-padia commented:
I've opened a PR at #29688
